modules/bcdec/image_decompress_bcdec.cpp

```diff
@@ -0,0 +1,180 @@
+		for (uint64_t x = 0; x < width; x += 4) {                                \
+			func(&src_blocks[src_pos], &dec_blocks[dst_pos], width *color_size); \
+			src_pos += block_size;                                               \
+			dst_pos += 4 * color_size;                                           \
+		}                                                                        \
+		dst_pos += 3 * width * color_size;                                       \
+	}
+
+	switch (format) {
+		case BCdec_BC1: {
+			DECOMPRESS_LOOP(bcdec_bc1, BCDEC_BC1_BLOCK_SIZE, 4)
+		} break;
+		case BCdec_BC2: {
+			DECOMPRESS_LOOP(bcdec_bc2, BCDEC_BC2_BLOCK_SIZE, 4)
+		} break;
+		case BCdec_BC3: {
+			DECOMPRESS_LOOP(bcdec_bc3, BCDEC_BC3_BLOCK_SIZE, 4)
+		} break;
+		case BCdec_BC4: {
+			DECOMPRESS_LOOP(bcdec_bc4, BCDEC_BC4_BLOCK_SIZE, 1)
+		} break;
+		case BCdec_BC5: {
+			DECOMPRESS_LOOP(bcdec_bc5, BCDEC_BC5_BLOCK_SIZE, 2)
+		} break;
+		case BCdec_BC6U: {
+			DECOMPRESS_LOOP(bcdec_bc6h_half_u, BCDEC_BC6H_BLOCK_SIZE, 6)
+		} break;
+		case BCdec_BC6S: {
+			DECOMPRESS_LOOP(bcdec_bc6h_half_s, BCDEC_BC6H_BLOCK_SIZE, 6)
+		} break;
+		case BCdec_BC7: {
+			DECOMPRESS_LOOP(bcdec_bc7, BCDEC_BC7_BLOCK_SIZE, 4)
+		} break;
+	}
+
+#undef DECOMPRESS_LOOP
+}
+
+void image_decompress_bcdec(Image *p_image) {
+	uint64_t start_time = OS::get_singleton()->get_ticks_msec();
+
+	int w = p_image->get_width();
+	int h = p_image->get_height();
+
+	Image::Format source_format = p_image->get_format();
+	Image::Format target_format = Image::FORMAT_MAX;
+
+	BCdecFormat dec_format = BCdec_BC1;
+
+	switch (source_format) {
+		case Image::FORMAT_DXT1:
+			dec_format = BCdec_BC1;
+			target_format = Image::FORMAT_RGBA8;
+			break;
+
+		case Image::FORMAT_DXT3:
+			dec_format = BCdec_BC2;
+			target_format = Image::FORMAT_RGBA8;
+			break;
+
+		case Image::FORMAT_DXT5:
+		case Image::FORMAT_DXT5_RA_AS_RG:
+			dec_format = BCdec_BC3;
+			target_format = Image::FORMAT_RGBA8;
+			break;
+
+		case Image::FORMAT_RGTC_R:
+			dec_format = BCdec_BC4;
+			target_format = Image::FORMAT_R8;
+			break;
+
+		case Image::FORMAT_RGTC_RG:
+			dec_format = BCdec_BC5;
+			target_format = Image::FORMAT_RG8;
+			break;
+
+		case Image::FORMAT_BPTC_RGBFU:
+			dec_format = BCdec_BC6U;
+			target_format = Image::FORMAT_RGBH;
+			break;
+
+		case Image::FORMAT_BPTC_RGBF:
+			dec_format = BCdec_BC6S;
+			target_format = Image::FORMAT_RGBH;
+			break;
+
+		case Image::FORMAT_BPTC_RGBA:
+			dec_format = BCdec_BC7;
+			target_format = Image::FORMAT_RGBA8;
+			break;
+
+		default:
+			ERR_FAIL_MSG("bcdec: Can't decompress unknown format: " + Image::get_format_name(source_format) + ".");
+			break;
+	}
+
+	Vector<uint8_t> data;
+	int64_t target_size = Image::get_image_data_size(w, h, target_format, p_image->has_mipmaps());
+	int mm_count = p_image->get_mipmap_count();
+	data.resize(target_size);
+
+	const uint8_t *rb = p_image->get_data().ptr();
+	uint8_t *wb = data.ptrw();
+
+	// Decompress mipmaps.
```

Comment:
@lyuma Can we check this? I've forgotten about how the mipmap texel squishing works.